server/cp/macros/usd_to_cad.py

```diff
@@ -0,0 +1,62 @@
+
+import re
+import logging
+import requests
+
+from superdesk.text_utils import get_text
+
+SERVICE_URL = 'https://www.bankofcanada.ca/valet/observations/FXUSDCAD/json?recent=1'
+
+CURRENCY_REGEX = re.compile(
+    r'(^| |\b)(?P<currency>[A-Z]+)?\$ ?(?P<num>[0-9]+(?P<decimal>\.[0-9]+)?)(?P<mil> million)?($|\b)',
+    re.MULTILINE,
+)
+
+sess = requests.Session()
+logger = logging.getLogger(__name__)
+
+
+def get_rate():
+    res = sess.get(SERVICE_URL, timeout=(3, 10))
+    res.raise_for_status()
+    rate = float(res.json()['observations'][0]['FXUSDCAD']['v'])
+    logger.debug('got USD2CAD rate %f', rate)
+    return rate
+
+
+def callback(item, **kwargs):
+    diff = {}
+    if not item.get('body_html'):
+        return diff
+    rate = get_rate()
+    text = get_text(item['body_html'], 'html', True)
+
+    def repl(m):
+        if m.group('currency') and m.group('currency') != 'US':
+            return
+        converted = float(m.group('num')) * rate
```

Comment:
does not really matter, but we can call `get_rate` after `if m.group('currency') and m.group('currency') != 'US':`

Reply:
that would mean calling it for every matching number instead of calling it once per macro call, so it might use different rates for each number..

Reply:
right, but we could make it called only once.. the idea is to call network only if match exists
anyway, it does not really matter

Reply:
yep I see your point, I would eventually use `@cache` for that `get_rate` if needed, I think they update the rate just once a day anyhow, but I don't think they will use it that much, so might be just complicating things